### PR TITLE
ceph: osd hide 'restorecon' when necessary

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

Some testing on OpenShift has shown that enabling Hostnetworking causes
issues with ceph-volume creating OSD.
This is a pure Red Hat OpenShift downstream issue.

This is meant to be temporary while we are waiting for the Selinux team
to get back to us.

**Which issue is resolved by this Pull Request:**
Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1764275

**Checklist:**

- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

### CODE REVIEW

File: pkg/daemon/ceph/osd/volume.go [hunk 1]
```diff
@@ -49,6 +50,11 @@ const (
 	dbDeviceFlag         = "--db-devices"
 	cephVolumeCmd        = "ceph-volume"
 	cephVolumeMinDBSize  = 1024 // 1GB
+	restoreConPath       = "/usr/sbin/restorecon"
+	restoreConNewContent = `#!/usr/bin/env bash
+echo "original restorecon command is now at %s"
```
Comment (travisn) [inline pkg/daemon/ceph/osd/volume.go]: Does this show up in the osd prepare log? How about a message that indicates restorecon was skipped?

File: pkg/daemon/ceph/osd/volume.go [hunk 2]
```diff
@@ -166,6 +176,39 @@ func updateLVMConfig(context *clusterd.Context, onPVC bool) error {
 	return nil
 }
 
+func hideRestoreconCommand(context *clusterd.Context) error {
```
Comment (travisn) [inline pkg/daemon/ceph/osd/volume.go]: should we add a unit test for this method?

File: pkg/daemon/ceph/osd/volume.go [hunk 3]
```diff
@@ -49,6 +50,11 @@ const (
 	dbDeviceFlag         = "--db-devices"
 	cephVolumeCmd        = "ceph-volume"
 	cephVolumeMinDBSize  = 1024 // 1GB
+	restoreConPath       = "/usr/sbin/restorecon"
```
Comment (travisn) [inline pkg/daemon/ceph/osd/volume.go]: If c-v thinks this command should be run, how can we just skip it with no consequences? The containers aren't affected?

**Comment by sp98** [comment]
`TestOSDAgenCephVolumeBluestore`  and `TestOSDAgenCephVolumeFilestore` unit tests are not failing.

**Comment by sp98** [comment]
> This needs more testing. DO NOT MERGE.

I have tested the changes in this branch with `hostNetwork` as `true` and downstream ceph image - `quay.io/rhceph-dev/rhceph-4.0-rhel-8:RHCEPH-4.0-RHEL-8-20190917.ci.0`.  The changes look good to me.
- Able to create OSDs with `hostNetwork`
- Able to update the OSD count. 
- Upgrading the `ceph` version also works.

File: pkg/daemon/ceph/osd/volume.go [hunk 4]
```diff
@@ -166,6 +176,39 @@ func updateLVMConfig(context *clusterd.Context, onPVC bool) error {
 	return nil
 }
 
+func hideRestoreconCommand(context *clusterd.Context) error {
+	// Check if Host Networking is enabled
+	hostnetworking := os.Getenv("ROOK_HOST_NETWORKING")
+	if hostnetworking == "false" {
+		logger.Debugf("ROOK_HOST_NETWORKING is %q, not hidding 'restorecon' command", hostnetworking)
```
[the remark this answers is not in the document]
Comment (sp98) [reply pkg/daemon/ceph/osd/volume.go]: looks more like a "replace" action.

File: pkg/daemon/ceph/osd/volume.go (hunk 3 above)
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go]: Right, c-v just runs the command and pursues if the exit code is 0. So this has no consequence.

File: pkg/daemon/ceph/osd/volume.go (hunk 1 above)
Comment (leseb) [reply pkg/daemon/ceph/osd/volume.go]: No, it only shows up in the `ceph-volume.log` file, let me clarify the message.

**Comment by alfredodeza** [comment]
Monkeypatching system executables at runtime might be problematic and packaging rules might even reject it. Can we get more clarification of what "temporary" means for this PR? For example, if the SELinux team doesn't get back to us in time, are we planning to release with this patch?

**Comment by leseb** [comment]
> Monkeypatching system executables at runtime might be problematic and packaging rules might even reject it. Can we get more clarification of what "temporary" means for this PR? For example, if the SELinux team doesn't get back to us in time, are we planning to release with this patch?

Yesterday, we all agreed that this is far from ideal. However, we rapidly need a solution to unblock testing. Currently, it is much faster to get this in Rook than a patch in c-v. In //, we should start a patch in c-v that should allow the ability not to run the `restorecon` command (via a flag). So far, packages rules haven't rejected that monkey-patching.
@alfredodeza, do you want to start a patch on c-v for this? Thanks.

**Comment by alfredodeza** [comment]
I understand this is trying to unblock testing. My question was to get some clarification of "temporary". If this is not to be released as-is for example, and just meant to unblock testing like you mention.

**Comment by leseb** [comment]
> I understand this is trying to unblock testing. My question was to get some clarification of "temporary". If this is not to be released as-is for example, and just meant to unblock testing like you mention.

You got it right, temporary means "once we have a patch in c-v downstream", this code will go away. We are not intending to have this as part of the release. The c-v patch could take different forms:

* one is the ability not to run any SELinux command (like restorecon)
* another is (depending on the SELinux team response) to drop all the SELinux related commands when running in containers

How does that sound?

**Comment by alfredodeza** [comment]
If the SELinux team says that we shouldn't run `restorecon` in this particular scenario then it would be sensible to get those changes into ceph-volume, I agree. Since this is allowing the testing to continue, I would prefer not to make changes in ceph-volume yet until we have confirmation on what is needed.

Thanks for clarifying @leseb

**Comment by obnoxxx** [comment]
@leseb, @alfredodeza, if I understand correctly, the recommendation of the selinux experts is not necessarily to not run `restorecon` at all, but to run it on the host instead of in the container. Am I getting something wrong here?

**Comment by leseb** [comment]
> @leseb, @alfredodeza, if I understand correctly, the recommendation of the selinux experts is not necessarily to not run `restorecon` at all, but to run it on the host instead of in the container. Am I getting something wrong here?

This is no recommendation from the selinux expert since no one replied yet. We decided not to run it in a container, we don't run it on the host either.

**Comment by obnoxxx** [comment]
> This is no recommendation from the selinux expert since no one replied yet.

@leseb  I thought we had received that recommendation in an internal email thread.

> We decided not to run it in a container, we don't run it on the host either.

Well, as the workaround, right. But I was commenting on the "real" fix that @alfredodeza talked about above.

**Pull request merged**